examples_tests/49.NewAPITest/main.cpp

```diff
@@ -0,0 +1,330 @@
+	{
+		video::IGPUGraphicsPipeline::SCreationParams gp_params;
+		gp_params.rasterizationSamplesHint = asset::IImage::ESCF_1_BIT;
+		gp_params.renderpass = renderpass;
+		gp_params.renderpassIndependent = rpindependent_pipeline;
+		gp_params.subpassIx = 0u;
+
+		pipeline = device->createGPUGraphicsPipeline(nullptr, std::move(gp_params));
+	}
+
+	core::smart_refctd_ptr<video::IGPUBuffer> buffer;
+	{
+		const SVertex vertices[3]{
+			{{0.0f, -0.5f}, {1.0f, 0.0f, 0.0f}},
+			{{0.5f, 0.5f}, {0.0f, 1.0f, 0.0f}},
+			{{-0.5f, 0.5f}, {0.0f, 0.0f, 1.0f}}
+		};
+
+		video::IDriverMemoryBacked::SDriverMemoryRequirements mreq;
+		
+		buffer = device->createDeviceLocalGPUBufferOnDedMem(sizeof(vertices));
+		assert(buffer);
+
+		core::smart_refctd_ptr<video::IGPUCommandBuffer> cb;
+		device->createCommandBuffers(cmdpool.get(), video::IGPUCommandBuffer::EL_PRIMARY, 1u, &cb);
+		assert(cb);
+
+		cb->begin(video::IGPUCommandBuffer::EU_ONE_TIME_SUBMIT_BIT);
+		cb->updateBuffer(buffer.get(), 0u, sizeof(vertices), vertices);
+		cb->end();
+		
+		video::IGPUQueue::SSubmitInfo info;
+		auto* cb_ = dynamic_cast<video::IGPUPrimaryCommandBuffer*>(cb.get());
+		info.commandBufferCount = 1u;
+		info.commandBuffers = &cb_;
+		info.pSignalSemaphores = nullptr;
+		info.signalSemaphoreCount = 0u;
+		info.pWaitSemaphores = nullptr;
+		info.waitSemaphoreCount = 0u;
+		info.pWaitDstStageMask = nullptr;
+		queue->submit(1u, &info, nullptr);
+	}
+
+	core::smart_refctd_ptr<video::IGPUCommandBuffer> cmdbuf[SC_IMG_COUNT];
+	device->createCommandBuffers(cmdpool.get(), video::IGPUCommandBuffer::EL_PRIMARY, SC_IMG_COUNT, cmdbuf);
+	for (uint32_t i = 0u; i < SC_IMG_COUNT; ++i)
+	{
+		auto& cb = cmdbuf[i];
+		auto& fb = fbo[i];
+
+		cb->begin(0);
+		
+		auto* buf = buffer.get();
+		size_t offset = 0u;
+		cb->bindVertexBuffers(0u, 1u, &buf, &offset);
+		cb->bindGraphicsPipeline(pipeline.get());
+		video::IGPUCommandBuffer::SRenderpassBeginInfo info;
+		asset::SClearValue clear;
+		asset::VkRect2D area;
+		area.offset = { 0, 0 };
+		area.extent = { WIN_W, WIN_H };
+		clear.color.float32[0] = 1.f;
+		clear.color.float32[1] = 0.f;
+		clear.color.float32[2] = 0.f;
+		clear.color.float32[0] = 1.f;
+		info.renderpass = renderpass;
+		info.framebuffer = fb;
+		info.clearValueCount = 1u;
+		info.clearValues = &clear;
+		info.renderArea = area;
+		cb->beginRenderPass(&info, asset::ESC_INLINE);
+		cb->draw(3u, 1u, 0u, 0u);
+		cb->endRenderPass();
+
+		cb->end();
+	}
+
+	constexpr uint32_t FRAME_COUNT = 5000u;
+	constexpr uint64_t MAX_TIMEOUT = 99999999999999ull; //ns
+	for (uint32_t i = 0u; i < FRAME_COUNT; ++i)
+	{
+		auto img_acq_sem = device->createSemaphore();
+		auto render_finished_sem = device->createSemaphore();
+
+		uint32_t imgnum = 0u;
+		sc->acquireNextImage(MAX_TIMEOUT, img_acq_sem.get(), nullptr, &imgnum);
+
+		video::IGPUQueue::SSubmitInfo submit;
+		{
+			auto* cb = dynamic_cast<video::IGPUPrimaryCommandBuffer*>(cmdbuf[imgnum].get());
```

Comment:
what's with the `dynamic_cast` ?

Reply:
IGPUCommandBuffer is virtual base, actually im thinking on getting rid of special class for primary command buffers because this tends to be annoying

Reply:
probably a good idea, no need to use OOP for differentiating between primary and secondary